### PR TITLE
ROC-7614 definition for issue general letter

### PR DESCRIPTION
**Pull request by akritis22**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/ROC-7614

### Change description ###
Please not this has been all reviewed as part of this PR https://github.com/hmcts/cmc-ccd-domain/pull/174
There is just one change made as part of this PR which is changing the new event to readonly (from CR to R) to prepare for release.

**Does this PR introduce a breaking change?**

```
- [ ] Yes
- [x ] No
```

### CODE REVIEW

**Comment by ckalista** [comment]
> Pls combine Letter related fields like Contact, Header, Prefix, RoleType(Claimant, Defendant), content, suffix into a complex type pls.

Oh yes, would be good to structure this as suggested 👍

File: definition/data/sheets/CaseField.json [hunk 1]
```diff
@@ -90,5 +90,7 @@
   {"CaseTypeID": "MoneyClaimCase", "ID": "directionOrder", "Label": "Direction order", "FieldType": "DirectionOrder", "SecurityClassification": "Public"},
   {"CaseTypeID": "MoneyClaimCase", "ID": "channel", "Label": "Channel", "FieldType": "FixedList", "FieldTypeParameter": "ChannelType", "SecurityClassification": "Public"},
   {"CaseTypeID": "MoneyClaimCase", "ID": "intentionToProceedDeadline", "Label": "Intention To Proceed Deadline", "FieldType": "Date", "FieldTypeParameter": "ChannelType", "SecurityClassification": "Public"},
-  {"CaseTypeID": "MoneyClaimCase", "ID": "state", "Label": "state", "FieldType": "Text", "SecurityClassification": "Public"}
+  {"CaseTypeID": "MoneyClaimCase", "ID": "state", "Label": "state", "FieldType": "Text", "SecurityClassification": "Public"},
+  {"CaseTypeID": "MoneyClaimCase", "ID": "generalLetterContent", "Label": "Who is the letter to?", "FieldType": "GeneralLetterContent", "SecurityClassification": "Public"},
```
Comment (karthikkannankainos) [inline definition/data/sheets/CaseField.json]: The label seems wrong. Pls check.

File: definition/data/sheets/ComplexTypes.json [hunk 2]
```diff
@@ -244,5 +244,13 @@
...
   {"ID": "ClaimSubmissionOperationIndicators", "ListElementCode": "sealedClaimUpload", "FieldType": "YesOrNo", "ElementLabel": "Upload of sealed claim successful", "SecurityClassification": "Public"},
   {"ID": "ClaimSubmissionOperationIndicators", "ListElementCode": "claimIssueReceiptUpload", "FieldType": "YesOrNo", "ElementLabel": "Upload of claim issue receipt successful", "SecurityClassification": "Public"},
-  {"ID": "ClaimSubmissionOperationIndicators", "ListElementCode": "defendantPinLetterUpload", "FieldType": "YesOrNo", "ElementLabel": "Upload of defendant pin letter successful", "SecurityClassification": "Public"}
+  {"ID": "ClaimSubmissionOperationIndicators", "ListElementCode": "defendantPinLetterUpload", "FieldType": "YesOrNo", "ElementLabel": "Upload of defendant pin letter successful", "SecurityClassification": "Public"},
+  {"ID": "GeneralLetterContent", "ListElementCode": "issueLetterContact", "ElementLabel": " ", "FieldType": "FixedRadioList", "FieldTypeParameter": "ChangeContactDetailsParty", "SecurityClassification": "Public"},
```
Comment (karthikkannankainos) [inline definition/data/sheets/ComplexTypes.json]: Element label missing. Pls verify.

File: definition/data/sheets/CaseField.json (hunk 1 above)
Comment (dharmendrak) [reply definition/data/sheets/CaseField.json]: they are not wrong , just to mimic the design the page

File: definition/data/sheets/ComplexTypes.json (hunk 2 above)
Comment (dharmendrak) [reply definition/data/sheets/ComplexTypes.json]: same as above

File: definition/data/sheets/AuthorisationCaseField.json [hunk 3]
```diff
@@ -267,5 +267,7 @@
   {"CaseTypeID": "MoneyClaimCase", "CaseFieldID": "expertReportInstruction", "UserRole": "caseworker-cmc-judge", "CRUD": "CRUD"},
   {"CaseTypeID": "MoneyClaimCase", "CaseFieldID": "dateReferredForDirections", "UserRole": "caseworker-cmc-judge", "CRUD": "CRUD"},
   {"CaseTypeID": "MoneyClaimCase", "CaseFieldID": "dateReferredForDirections", "UserRole": "citizen", "CRUD": "CRUD"},
-  {"CaseTypeID": "MoneyClaimCase", "CaseFieldID": "dateReferredForDirections", "UserRole": "caseworker-cmc", "CRUD": "CRUD"}
+  {"CaseTypeID": "MoneyClaimCase", "CaseFieldID": "dateReferredForDirections", "UserRole": "caseworker-cmc", "CRUD": "CRUD"},
+  {"CaseTypeID": "MoneyClaimCase", "CaseFieldID": "generalLetterContent", "UserRole": "caseworker-cmc", "CRUD": "CRUD"},
```
Comment (ckalista) [inline definition/data/sheets/AuthorisationCaseField.json]: Should we not make these read-only, or else the event will show in production before the actual release date?

File: definition/data/sheets/AuthorisationCaseField.json [hunk 4]
```diff
@@ -267,5 +267,7 @@
...
+  {"CaseTypeID": "MoneyClaimCase", "CaseFieldID": "draftLetterDoc", "UserRole": "caseworker-cmc", "CRUD": "CRUD"}
```
Comment (ckalista) [inline definition/data/sheets/AuthorisationCaseField.json]: Same comment as above.

File: definition/data/sheets/CaseEvent.json [hunk 5]
```diff
@@ -66,5 +66,6 @@
...
+  {"CaseTypeID": "MoneyClaimCase", "ID": "ResetPin", "Name": "Reset Pin", "Description": "Reset and send pin to defendant", "DisplayOrder": 1,"PreConditionState(s)": "*","PostConditionState": "*", "SecurityClassification": "Public", "CallBackURLAboutToSubmitEvent": "${CCD_DEF_CLAIM_STORE_BASE_URL}/cases/callbacks/about-to-submit", "RetriesTimeoutAboutToSubmitEvent": "5,10,15"},
+  {"CaseTypeID": "MoneyClaimCase", "ID": "IssueGeneralLetter", "Name": "Issue General Letter", "Description": "Issue a letter response to any party", "DisplayOrder": 1,"PreConditionState(s)": "*","PostConditionState": "*", "SecurityClassification": "Public", "CallBackURLAboutToStartEvent": "${CCD_DEF_CLAIM_STORE_BASE_URL}/cases/callbacks/about-to-start", "RetriesTimeoutAboutToStartEvent": "5,10,15", "CallBackURLAboutToSubmitEvent": "${CCD_DEF_CLAIM_STORE_BASE_URL}/cases/callbacks/about-to-submit", "RetriesTimeoutAboutToSubmitEvent": "5,10,15"}
```
Comment (ckalista) [inline definition/data/sheets/CaseEvent.json]: We may want to have a quick chat with @dharmendrak about this retry strategy. Not sure we should use it as could lead to side-effect...

File: definition/data/sheets/CaseEvent.json [hunk 6]
```diff
@@ -66,5 +66,6 @@
...
-  {"CaseTypeID": "MoneyClaimCase", "ID": "ResetPin", "Name": "Reset Pin", "Description": "Reset and send pin to defendant", "DisplayOrder": 1,"PreConditionState(s)": "*","PostConditionState": "*", "SecurityClassification": "Public", "CallBackURLAboutToSubmitEvent": "${CCD_DEF_CLAIM_STORE_BASE_URL}/cases/callbacks/about-to-submit", "RetriesTimeoutAboutToSubmitEvent": "5,10,15"}
+  {"CaseTypeID": "MoneyClaimCase", "ID": "ResetPin", "Name": "Reset Pin", "Description": "Reset and send pin to defendant", "DisplayOrder": 1,"PreConditionState(s)": "*","PostConditionState": "*", "SecurityClassification": "Public", "CallBackURLAboutToSubmitEvent": "${CCD_DEF_CLAIM_STORE_BASE_URL}/cases/callbacks/about-to-submit", "RetriesTimeoutAboutToSubmitEvent": "5,10,15"},
```
Comment (ckalista) [inline definition/data/sheets/CaseEvent.json]: We may want to have a quick chat with @dharmendrak about this retry strategy. Not sure we should use it as could lead to side-effect...

File: definition/data/sheets/CaseEventToFields.json [hunk 7]
```diff
@@ -70,6 +70,7 @@
...
   {"CaseTypeID": "MoneyClaimCase", "CaseEventID": "SuccessfulMediation", "CaseFieldID": "respondents", "PageID": 1, "PageDisplayOrder": 1, "DisplayContext": "COMPLEX", "PageFieldDisplayOrder": 1},
   {"CaseTypeID": "MoneyClaimCase", "CaseEventID": "SuccessfulMediation", "CaseFieldID": "staffUploadedDocuments", "PageID": 2, "PageDisplayOrder": 2,"DisplayContext": "COMPLEX", "PageLabel":"Upload mediation agreement","PageFieldDisplayOrder": 1},
-  {"CaseTypeID": "MoneyClaimCase", "CaseEventID": "FailedMediation", "CaseFieldID": "respondents", "PageID": 1, "PageDisplayOrder": 1, "DisplayContext": "COMPLEX", "PageFieldDisplayOrder": 1}
-
+  {"CaseTypeID": "MoneyClaimCase", "CaseEventID": "FailedMediation", "CaseFieldID": "respondents", "PageID": 1, "PageDisplayOrder": 1, "DisplayContext": "COMPLEX", "PageFieldDisplayOrder": 1},
+  {"CaseTypeID": "MoneyClaimCase", "CaseEventID": "IssueGeneralLetter", "CaseFieldID": "generalLetterContent", "PageID": 1, "PageDisplayOrder": 1, "DisplayContext": "COMPLEX", "PageFieldDisplayOrder": 1, "CallBackURLMidEvent": "${CCD_DEF_CLAIM_STORE_BASE_URL}/cases/callbacks/mid", "RetriesTimeoutURLMidEvent": "5,10,15"},
```
Comment (ckalista) [inline definition/data/sheets/CaseEventToFields.json]: Same here - retry strategies need to be challenged.

File: definition/data/sheets/AuthorisationCaseField.json (hunk 3 above)
Comment (akritis22) [reply definition/data/sheets/AuthorisationCaseField.json]: Yes Chris they will be read only..just changed for time being to regression test to make sure nothing breaks after the review comments are addressed.

File: definition/data/sheets/CaseEventToFields.json (hunk 7 above)
Comment (akritis22) [reply definition/data/sheets/CaseEventToFields.json]: do you suggest not having any retries?

**Pull request merged**